### PR TITLE
LibX86+disasm: Support x86-64

### PR DESCRIPTION
**Pull request by skyrising**

### CODE REVIEW

File: Userland/Libraries/LibX86/Instruction.h [hunk 1]
```diff
@@ -202,6 +203,7 @@ enum InstructionFormat {
 };
 
 static constexpr unsigned CurrentAddressSize = 0xB33FBABE;
+static constexpr unsigned CurrentOperandSize = 0xB33FB00F;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.h:211]: This seems duplicate

File: Userland/Libraries/LibX86/Disassembler.h [hunk 2]
```diff
@@ -25,8 +25,12 @@ class Disassembler {
 #if ARCH(I386)
         return Instruction::from_stream(m_stream, OperandSize::Size32, AddressSize::Size32);
 #else
-        dbgln("FIXME: Implement disassembly support for x86_64");
+#    if ARCH(X86_64)
+        return Instruction::from_stream(m_stream, OperandSize::Size64, AddressSize::Size64);
```
Comment (IdanHo) [inline Userland/Libraries/LibX86/Disassembler.h:29]: The default operand size in long mode is 32 bit

File: Userland/Utilities/disasm.cpp [hunk 3]
```diff
@@ -130,7 +130,26 @@ ErrorOr<int> serenity_main(Main::Arguments args)
...
+        builder.appendff("{:p}  ", virtual_offset);
+        for (size_t i = 0; i < 7; i++) {
+            if (i < length)
+                builder.appendff("{:02x} ", asm_data[offset + i]);
+            else
+                builder.append("   "sv);
+        }
+        builder.append(" "sv);
+        builder.append(insn.value().to_string(virtual_offset, symbol_provider));
+        outln("{}", builder.to_string());
```
Comment (awesomekling) [inline Userland/Utilities/disasm.cpp:145]: You can avoid a string allocation here by outputting `builder.string_view()` instead of calling `to_string()`.

cc @nico :^)

File: Userland/Libraries/LibX86/Instruction.cpp [hunk 4]
```diff
@@ -1084,11 +1093,57 @@ static void build_sse_f3(u8 op, const char* mnemonic, InstructionFormat format,
...
     build_0f(0xFE, "PADDD", OP_mm1_mm2m64, &Interpreter::PADDD_mm1_mm2m64);
     build_0f(0xFF, "UD0", OP, &Interpreter::UD0);
+
+    auto* table64 = s_table[to_underlying(OperandSize::Size64)];
+    table64[0x06] = {};
+    table64[0x07] = {};
+    table64[0x16] = {};
+    table64[0x17] = {};
+    table64[0x27] = {};
+    table64[0x37] = {};
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.cpp:1103]: What are these?

File: Userland/Libraries/LibX86/Instruction.h [hunk 5]
```diff
@@ -510,8 +596,12 @@ class MemoryOrRegisterReference {
...
     };
 
-    u8 m_rm_byte { 0 };
-    u8 m_sib { 0 };
+    u8 m_mod : 2 { 0 };
+    u8 m_reg : 4 { 0 };
+    u8 m_rm : 4 { 0 };
+    u8 m_sib_scale : 2 { 0 };
+    u8 m_sib_index : 4 { 0 };
+    u8 m_sib_base : 4 { 0 };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.h:604]: can you factor out these changes into two separate commits?

File: Userland/Libraries/LibX86/Instruction.h [hunk 6]
```diff
@@ -901,6 +999,11 @@ ALWAYS_INLINE Instruction::Instruction(InstructionStreamType& stream, OperandSiz
     : m_operand_size(operand_size)
     , m_address_size(address_size)
 {
+    if (operand_size == OperandSize::Size64 || address_size == AddressSize::Size64) {
+        m_operand_size = OperandSize::Size32;
+        m_address_size = AddressSize::Size32;
```
Comment (IdanHo) [inline Userland/Libraries/LibX86/Instruction.h:1004]: The default address_size on x86_64 is 64 bits

**Review by IdanHo — COMMENTED** [review]
You probably already accounted for these 2 in some way in the code, but we might want to change this to be consistent with the intel manual.

**Review by Hendiadyoin1 — COMMENTED** [review]
Nice work!
I haven't read it with my keenest eye yet, but it looks mostly good

Some general stuff would be, that more fine grained commits would be nice, and would ease review
And also you should maybe add compile time guards around the x64 specific code

File: Userland/Libraries/LibX86/Instruction.cpp [hunk 7]
```diff
@@ -1336,6 +1338,62 @@ static String sib_to_string(u8 rm, u8 sib)
...
+
+    if (!has_displacement)
+        return base;
+
+    String displacement_string;
+    if ((i32)m_displacement32 < 0)
+        displacement_string = String::formatted("-{:#x}", -(i32)m_displacement32);
+    else
+        displacement_string = String::formatted("+{:#x}", m_displacement32);
+    return String::formatted("{}{}", base, displacement_string);
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.cpp:1406]: cant we just always cast?
String::format should have a setting to always print the sign

File: Userland/Libraries/LibX86/Instruction.cpp [hunk 8]
```diff
@@ -1343,46 +1366,33 @@ static String sib_to_string(u8 rm, u8 sib)
...
 {
     if (is_register())
-        return register_name(static_cast<RegisterIndex32>(m_register_index));
+        return register_name(static_cast<RegisterIndex64>(m_register_index));
 
     bool has_displacement = false;
     switch (mod()) {
+    case 0b00:
+        has_displacement = m_rm == 5;
+        break;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.cpp:1375]: Didn't you just add this function in one of the previous commits,
Maybe fix it up there.

File: Userland/Libraries/LibX86/Instruction.cpp [hunk 9]
```diff
@@ -1252,79 +1319,35 @@ String MemoryOrRegisterReference::to_string_a16() const
     return String::formatted("{}{}", base, displacement_string);
 }
 
-static String sib_to_string(u8 rm, u8 sib)
+String MemoryOrRegisterReference::sib_to_string(bool long_mode) const
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.cpp:1322]: Maybe use a enum for the argument and not a magic bool

File: Userland/Libraries/LibX86/Instruction.h [hunk 10]
```diff
@@ -213,17 +218,19 @@ struct InstructionDescriptor {
     bool has_rm { false };
     unsigned imm1_bytes { 0 };
     unsigned imm2_bytes { 0 };
+    bool long_mode_default_64 { false };
+    bool long_mode_force_64 { false };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.h:222]: We dont expect to have any 32 bit binaries when using the x64 build, so you can elide any full backwards compatibility features

Confirmation by maintainer needed

File: Userland/Libraries/LibX86/Instruction.h [hunk 11]
```diff
@@ -619,9 +712,14 @@ class Instruction {
...
     OperandSize m_operand_size { OperandSize::Size16 };
     AddressSize m_address_size { AddressSize::Size16 };
+    ProcessorMode m_mode { ProcessorMode::Protected };
     bool m_has_lock_prefix : 1 { false };
     bool m_has_operand_size_override_prefix : 1 { false };
     bool m_has_address_size_override_prefix : 1 { false };
+    bool m_has_rex_w : 1 { false };
+    bool m_has_rex_r : 1 { false };
+    bool m_has_rex_x : 1 { false };
+    bool m_has_rex_b : 1 { false };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.h:722]: Maybe put these next to the modrm and sib bytes, because the REX prefix seems to act similar to these
than just a `m_has_rex_refix` flag would remain, but haven't read that part of the manual fully

Also using an anonymous unions might ease initialization and overriding of this (same for modrm/sib)

E: Looking further up/down it seems like you have also extended the modrm/sib parts to account for these there so a union would not work anymore....

**Review by Hendiadyoin1 — COMMENTED** [review]
With a more thorough reading I found some more things

File: Userland/Libraries/LibX86/Instruction.h (hunk 10 above)
Comment (skyrising) [reply Userland/Libraries/LibX86/Instruction.h:222]: Yeah, at the moment disasm assumes it only gets ELF files for the platform it's running on. But maybe we want to make it take that information from the ELF header in the future?

File: Userland/Libraries/LibX86/Instruction.h (hunk 11 above)
Comment (skyrising) [reply Userland/Libraries/LibX86/Instruction.h:722]: I experimented with a union now and I don't really like how it would turn out. I'd prefer keeping these bools.

File: Userland/Libraries/LibX86/Instruction.h (hunk 1 above)
Comment (skyrising) [reply Userland/Libraries/LibX86/Instruction.h:211]: `CurrentAddressSize` is the magic number to use the address size as immediate size (used by moffset and near branches). `CurrentOperandSize` is for using the operand size (used by movsxd and affected by rex.w)

Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Instruction.h:211]: Oh I can't read...

File: Userland/Libraries/LibX86/Instruction.cpp (hunk 7 above)
Comment (alimpfard) [reply Userland/Libraries/LibX86/Instruction.cpp]: `:+#x`
